dask/_expr.py

```diff
@@ -733,3 +757,236 @@ def collect_dependents(expr) -> defaultdict:
             stack.append(dep)
             dependents[dep._name].append(weakref.ref(node))
     return dependents
+
+
+class HLGExpr(Expr):
+    _parameters = [
+        "dsk",
+        "low_level_optimizer",
+        "output_keys",
+        "finalize_compute",
+    ]
+    _defaults = {
+        "low_level_optimizer": None,
+        "output_keys": None,
+        "finalize_compute": None,
+    }
+
+    @staticmethod
+    def from_collection(collection, optimize_graph=True):
+        if hasattr(collection, "dask"):
+            dsk = collection.dask.copy()
+        else:
+            dsk = collection.__dask_graph__()
+            from dask.highlevelgraph import HighLevelGraph
+
+            if not isinstance(dsk, HighLevelGraph):
+
+                dsk = HighLevelGraph.from_collections(
+                    str(id(collection)), dsk, dependencies=()
+                )
+        return HLGExpr(
+            dsk=dsk,
+            low_level_optimizer=(
+                collection.__dask_optimize__ if optimize_graph else None
+            ),
+            output_keys=collection.__dask_keys__(),
+            finalize_compute=collection.__dask_postcompute__,
+        )
+
+    def __dask_annotations__(self):
+        annotations_by_type = {}
+        for layer in self.operand("dsk").layers.values():
+            if layer.annotations:
+                annot = layer.annotations
+                for annot_type, value in annot.items():
+                    annotations_by_type[annot_type].update(
+                        {k: (value(k) if callable(value) else value) for k in layer}
+                    )
+        return annotations_by_type
+
+    def _operands_for_repr(self):
+        return [
+            f"name={self.operand('name')!r}",
+            f"dsk={self.operand('dsk')!r}",
+        ]
+
+    def _tree_repr_lines(self, indent=0, recursive=True):
+        # TODO: This doesn't make sense yet
+        return self._operands_for_repr()
+
+    def __dask_keys__(self):
+        if self.operand("output_keys"):
+            return self.operand("output_keys")
+        dsk = self.operand("dsk")
+        # Note: This will materialize
+        dependencies = dsk.get_all_dependencies()
+        dependents = reverse_dict(dependencies)
+        return [d for d in dependents if not dependents[d] and d in dsk]
+
+    def _layer(self) -> dict:
+        dsk = self.operand("dsk")
+
+        keys = self.operand("output_keys")
+        optimizer = self.operand("low_level_optimizer")
+        if keys is None and optimizer is not None:
+            keys = self.__dask_keys__()
+
+        if (optimizer := self.operand("low_level_optimizer")) is not None:
+            return ensure_dict(optimizer(dsk, keys))
+        return ensure_dict(dsk)
+
+
+class HLGTuple(Expr):
+    _parameters = ["hlgs", "optimizer"]
```

Comment:
Not sure, but this seems odd. We normally don't pass expressions in as lists, rather as *args. But maybe that's fine for HLGs?

Reply:
yeah, this felt odd with having another argument `optimizer` but should no be a big deal either way. I can change that.

Reply:
ok I believe the only way to do a *args AND pass another argument is to just not define _parameters at all which makes param lookup difficult. Without a proper `__init__` I don't think we can do this.